cirq/study/sweep_trial_result.py

```diff
@@ -0,0 +1,97 @@
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Defines sweep trial results."""
+
+from typing import Dict, Any, TYPE_CHECKING, List, Set
+
+from cirq import value
+from cirq.study import resolver, trial_result
+
+if TYPE_CHECKING:
+    # pylint: disable=unused-import
+    import cirq
+
+
+@value.value_equality(unhashable=True)
+class SweepTrialResult:
+    """The combination of multiple TrialResults with potentially different
+       set of parameters.
+
+    Attributes:
+        trial_results: A list of TrialResults.
+    """
+
+    def __init__(self,
+                 trial_results: List[trial_result.TrialResult] = None) -> None:
+        self.trial_results = trial_results if trial_results is not None else []
+        self._index = {}  # type: Dict[str, Set[int]]
+        for i, result in enumerate(self.trial_results):
+            for key, value in result.params.param_dict.items():
+                self._index.setdefault(key + str(value), set()).add(i)
+
+    def trials_where_params_match(
+            self, param_dict: resolver.ParamResolverOrSimilarType
+    ) -> List[trial_result.TrialResult]:
+        """
+        Gets the list of TrialResults that match the input mapping of
+        parameters to their values. Takes an "and" of all the specified
+        conditions.
+
+        Args:
+            param_dict: A dictionary from the ParameterValue key (str) to its
+                assigned value.
+        Returns:
+            A list of TrialResults.
+        """
+        trial_results_match = set(range(len(self.trial_results)))
+        for key, value in resolver.ParamResolver(param_dict).param_dict.items():
+            trial_results_match.intersection_update(
+                self._index.get(key + str(value), set()))
+        return [self.trial_results[i] for i in trial_results_match]
+
+    def slice_where_params_match(self,
+                                 param_dict: resolver.ParamResolverOrSimilarType
+                                ) -> 'cirq.SweepTrialResult':
+        """
+        Gets a SweepTrialResult which represents a slice of the current one
+        limited to the TrialResults that match the input mapping of parameters
+        to their values. Takes an "and" of all the specified conditions.
+
+        Args:
+            param_dict: A dictionary from the ParameterValue key (str) to its
+                assigned value.
+        Returns:
+            A SweepTrialResult limited to TrialResults that match the specified
+            input.
+        """
+        return SweepTrialResult(self.trials_where_params_match(param_dict))
+
```

Comment:
Implement `__iter__`, `__getitem__` and `__len__` so that this acts a bit more like a list of trial results.